README.md
# fuzzy-sql

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2019 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-07-03
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-07-03
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/fuzzy-sql/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/fuzzy-sql/compare/0.1.0...0.1.1


test/fuzzy_sql/core_test.clj
(ns fuzzy-sql.core-test
  (:require [clojure.test :refer :all]
            [fuzzy-sql.exporter :refer :all]
            [next.jdbc :as jdbc])
  (:import (java.util UUID)))



(def test-data [["header" "age" "someNulls"]
                ["value 1" 3 nil]
                ["value 2" 4 3.4]])


(def db {:dbtype   "mysql"
         :dbname   "dsd"
         :user     "root"
         :password "pass"
         :host     "localhost"
         :useSSl   false})

(def ds (jdbc/get-datasource db))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 1 1))))

(deftest test-index-of
  (let [data [:a :b :c :d]
        c-index (index-of #(= :c %1) data)]
    (is (= c-index 2))))

(deftest test-detect-data-type
  (testing "detecting data type"
    (is (= (detect-data-type test-data "someNulls") Double))
    (is (= (detect-data-type test-data "header") String))
    (is (= (detect-data-type test-data "age") Long))))


(deftest test-detect-db-type
  (testing "detecting data type"
    (is (= (detect-db-type test-data "someNulls") "numeric(19,6)"))
    (is (= (detect-db-type test-data "header") "varchar(255)"))
    (is (= (detect-db-type test-data "age") "bigint"))))


(deftest test-generate-ddl
  (testing "testing generate ddl"
    (is (= (generate-ddl test-data "my_table")
           "create table `my_table` (`header` varchar(255),`age` bigint,`someNulls` numeric(19,6))"))))

(deftest test-insertion
  (testing "creating the table"
    (let [rand-tb-name (-> (UUID/randomUUID) str (.replace "-" ""))]
      (println "inserting in table" rand-tb-name)
      (create-or-insert ds test-data rand-tb-name))))








src/fuzzy_sql/db.clj
(ns fuzzy-sql.db
  (:require [next.jdbc :as jdbc]
            [next.jdbc.result-set :as rs]))


(defn get-columns [ds table]
  (with-open [c (jdbc/get-connection ds)]
    (-> (.getMetaData c)
        (.getColumns nil nil table nil)
        (rs/datafiable-result-set ds {}))))


(defn get-tables [ds table]
  (with-open [c (jdbc/get-connection ds)]
    (-> (.getMetaData c)
        (.getTables nil nil table (into-array String ["TABLE"]))
        (rs/datafiable-result-set ds {}))))



;[{:IS_NULLABLE "YES",
;  :IS_GENERATEDCOLUMN "NO",
;  :COLUMN_NAME "header",
;  :SQL_DATA_TYPE 0,
;  :COLUMN_DEF nil,
;  :TABLE_CAT "dsd",
;  :SCOPE_CATALOG nil,
;  :SQL_DATETIME_SUB 0,
;  :DECIMAL_DIGITS nil,
;  :COLUMN_SIZE 255,
;  :REMARKS "",
;  :BUFFER_LENGTH 65535,
;  :TYPE_NAME "VARCHAR",
;  :ORDINAL_POSITION 1,
;  :SOURCE_DATA_TYPE nil,
;  :NUM_PREC_RADIX 10,
;  :CHAR_OCTET_LENGTH 255,
;  :SCOPE_SCHEMA nil,
;  :SCOPE_TABLE nil,
;  :NULLABLE 1,
;  :TABLE_SCHEM nil,
;  :IS_AUTOINCREMENT "NO",
;  :TABLE_NAME "table1",
;  :DATA_TYPE 12}
; {:IS_NULLABLE "YES",
;  :IS_GENERATEDCOLUMN "NO",
;  :COLUMN_NAME "age",
;  :SQL_DATA_TYPE 0,
;  :COLUMN_DEF nil,
;  :TABLE_CAT "dsd",
;  :SCOPE_CATALOG nil,
;  :SQL_DATETIME_SUB 0,
;  :DECIMAL_DIGITS 0,
;  :COLUMN_SIZE 19,
;  :REMARKS "",
;  :BUFFER_LENGTH 65535,
;  :TYPE_NAME "BIGINT",
;  :ORDINAL_POSITION 2,
;  :SOURCE_DATA_TYPE nil,
;  :NUM_PREC_RADIX 10,
;  :CHAR_OCTET_LENGTH nil,
;  :SCOPE_SCHEMA nil,
;  :SCOPE_TABLE nil,
;  :NULLABLE 1,
;  :TABLE_SCHEM nil,
;  :IS_AUTOINCREMENT "NO",
;  :TABLE_NAME "table1",
;  :DATA_TYPE -5}
; {:IS_NULLABLE "YES",
;  :IS_GENERATEDCOLUMN "NO",
;  :COLUMN_NAME "someNulls",
;  :SQL_DATA_TYPE 0,
;  :COLUMN_DEF nil,
;  :TABLE_CAT "dsd",
;  :SCOPE_CATALOG nil,
;  :SQL_DATETIME_SUB 0,
;  :DECIMAL_DIGITS 6,
;  :COLUMN_SIZE 19,
;  :REMARKS "",
;  :BUFFER_LENGTH 65535,
;  :TYPE_NAME "DECIMAL",
;  :ORDINAL_POSITION 3,
;  :SOURCE_DATA_TYPE nil,
;  :NUM_PREC_RADIX 10,
;  :CHAR_OCTET_LENGTH nil,
;  :SCOPE_SCHEMA nil,
;  :SCOPE_TABLE nil,
;  :NULLABLE 1,
;  :TABLE_SCHEM nil,
;  :IS_AUTOINCREMENT "NO",
;  :TABLE_NAME "table1",
;  :DATA_TYPE 3}]

src/fuzzy_sql/countdown.clj
(def data (loop [arr3 [], x 5]
            (if (zero? x)
              arr3
              (recur (conj arr3 x) (dec x)))))


(defn last-index [vec] (- (count vec) 1))

(defn head [vector] (subvec vector 0 (last-index vector)))

(defn recursive-reverse [coll]
  (loop [result [],
         x coll]
    (if (= (count x) 0)
      result
      (recur (conj result (last x)) (head x)))))

(println "recursive reverse")
(recursive-reverse [1, 2, 3])

; =====================================================
; Index of any
;===============================================

(defn indexed [coll] (map-indexed vector coll))

(defn index-filter [pred coll]
  (when pred
    (for [[idx elt] (indexed coll) :when (pred elt)] idx)))


(index-filter #{\a \b} "abcdbbb")

(nth (index-filter #{:h} [:t :t :h :t :h :t :t :t :h :h]) 2)

(meta #'iterate)

(class (rest '(1 2 3)))
;===========
;list comprehension

(defn whole-numbers [] (iterate inc 1))

(for [word ["the" "quick" "brown" "fox"]]
  (format "<p>%s</p>" word))


(take 10 (for [n (whole-numbers) :when (even? n)] n))

; interesting fibonacci
(take 5 (iterate (fn [[a b]] [b (+ a b)]) [0 1]))

src/fuzzy_sql/exporter.clj
(ns fuzzy-sql.exporter
  (:require
    [fuzzy-sql.db :as db]
    [next.jdbc :as jdbc]
    [next.jdbc.sql :as sql])
  (:import (java.util Date)))

(declare unwind-values-to-keys)

(defn create-map-vk [[k values]]
  (if (coll? values)
    (reduce (fn [a v] (assoc a v k)) {} values)
    {values k}))

(defn unwind-values-to-keys [data-map]
  (reduce (fn [acc v] (into (create-map-vk v) acc)) {} data-map))


(def MYSQL-TO-JAVA-MAPPINGS {"varchar(255)"  [String]
                             "bigint"        [Integer Long BigInteger]
                             "numeric(19,6)" [Float Double BigDecimal]
                             "datetime"      [java.sql.Date Date]})


(def JAVA-TO-SQL-DB-MAPPING (unwind-values-to-keys MYSQL-TO-JAVA-MAPPINGS))


(defn filter-and-pair-with-index [pred? coll]
  (let [data-and-index (map-indexed vector coll)
        filtered-item-pairs (filter (fn [[_ item]] (pred? item)) data-and-index)]
    filtered-item-pairs))


(defn index-of [pred? coll]
  (let [[first-pair] (filter-and-pair-with-index pred? coll)
        [index] first-pair] index))


(defn get-header [csv] (first csv))


(defn get-header-index [csv name]
  (let [header (get-header csv)
        index (index-of #(= %1 name) header)] index))

(defn get-data [csv] (subvec csv 1 (count csv)))


(defn get-column [csv col-name]
  (let [index (get-header-index csv col-name)
        data (get-data csv)
        col-values (map #(nth %1 index) data)] col-values))


(defn detect-data-type [csv name]
  (let [col-values (get-column csv name)
        first-value (first (filter some? col-values))]
    (type first-value)))


(defn detect-db-type [csv name] (JAVA-TO-SQL-DB-MAPPING (detect-data-type csv name)))

(defn coalesce [a b] (if (nil? a) b a))

(defn generate-ddl [csv name]
  (let [headers (get-header csv)
        col-expr (map #(str "`" %1 "` " (detect-db-type csv %1)) headers)
        all-col-exprs (reduce #(str %1 "," %2) col-expr)]
    (str "create table `" name "` (" all-col-exprs ")")))

(defn to-map-list [csv]
  (let [[header & data] csv
        records (map #(zipmap header %1) data)]
    records))


(defn weigh [v]
  (cond
    (instance? String v) {:size (count v) :decimals 0}
    (instance? Number v) (let [[size decimals] (.split (str v) "\\.")]
                           {:size (count size) :decimals (count (coalesce decimals ""))})
    (instance? Date v) {:size 1 :decimals 0}
    :else {:size (count (str v)) :decimals 0}))


(defn- sum-weight [weight] (+ (:size weight) (:decimals weight)))


(defn detect-max-size-per-header [csv]
  (let [header (get-header csv)
        largest-per-header (map #(apply max
                                        (map sum-weight
                                             (map weigh (get-column csv %1))))
                                header)
        header-and-max (zipmap header largest-per-header)]
    header-and-max))


(defn create-table [ds csv name]
  (let [ddl-string (generate-ddl csv name)]
    (jdbc/execute! ds [ddl-string])))

(defn insert-data [ds csv name]
  (let [[header & data] csv
        table-name (keyword name)
        csv-header-key (map #(keyword %1) header)]
    (sql/insert-multi! ds table-name csv-header-key data)))



(defn create-or-insert [ds csv name]
  (let [table-record (db/get-tables ds name)
        table-count (count table-record)]
    (when (= 0 table-count) (create-table ds csv name))
    (insert-data ds csv name)))



(defn resize-if-necessary [ds csv table-name]
  (let [header-sizes (detect-max-size-per-header csv)]
    header-sizes))




; check if table exists
; create table
; for each record insert
; if fails to to adjust table
; insert again

